llvm/lib/Transforms/Vectorize/LoopVectorize.cpp

```diff
@@ -9019,6 +9152,15 @@ LoopVectorizationPlanner::tryToBuildVPlanWithVPRecipes(VFRange &Range) {
         bool NeedsBlends = BB != HeaderBB && !BB->phis().empty();
         return Legal->blockNeedsPredication(BB) || NeedsBlends;
       });
+
+  // Cache the partial reductions up front so we can remove the invalid ones
+  // before creating the recipes
+  for (const auto &[Phi, RdxDesc] : Legal->getReductionVars())
+    if (std::optional<PartialReductionChain> Chain =
+            getScaledReduction(Phi, RdxDesc, &TTI, Range, CM))
+      RecipeBuilder.addScaledReductionExitInstr(*Chain);
+  RecipeBuilder.removeInvalidScaledReductionExitInstrs();
```

Comment:
If those are only used together, might be better to keep them together instead of exposing them separately publicly in the API

Reply:
The issue is that we need to have all partial reductions populated before we can remove the invalid ones so `addScaledReductionExitInstr` and `removeInvalidScaledReductionExitInstrs` need to be separate.

Reply:
I think you could add Chain instances to a local SmallVector here, then pass that to a single interface in RecipeBuilder which filtered the results and only stored it to the ScaledReductionExitInstrs map if there were no additional uses.

Reply:
Ah I like that idea Graham, since it is a bit more efficient than adding then removing. I've tried that approach but will keep Florian's suggestion in mind if you don't like how it's been done.